### DISCUSSION TITLE
How to make menus?

### DISCUSSION THREAD

**Discussion by The-Enazer** [Techokami/SonicWorldsNext#207] [Q&A] [answered]
hello it's me again the guy of the data select. i come again for another question. How to make a menu? (I tried to make it but dont work) Plz help (sorry for the bad english)

**Comment by r6915ee** [comment]
Most of it should conduct of the same techniques you would use to make a regular menu in Godot. As for some of the specific things that Worlds Next provides, you can make the menus change to other menus and the game as you normally would.

**Reply by The-Enazer** [reply]
I tried that, i use the control node and the button node but dont work.

**Reply by r6915ee** [reply]
Did you connect the necessary signals and add the code that reacts to each signal?

**Reply by The-Enazer** [reply]
Yeah, the button no are clickable and i put a code that i investigate to use the keyboard and nothing

**Reply by r6915ee** [reply]
Can you perhaps show a screenshot or video of the menu issues occurring?

**Reply by The-Enazer** [reply]
![Captura desde 2024-10-17 23-09-52](https://github.com/user-attachments/assets/d262dbfb-1fb7-4871-86ea-55510d5e2995)
![Captura desde 2024-10-17 23-09-59](https://github.com/user-attachments/assets/89be93ea-d168-4d94-a65a-30f616a17642)

**Reply by The-Enazer** [reply]
lmao

**Reply by The-Enazer** [reply]
(note: sorry if the code is bad i am no the best in gdscript)

**Reply by r6915ee** [reply, marked as answer]
> ![Captura desde 2024-10-17 23-09-59](https://private-user-images.githubusercontent.com/179539598/377694920-89be93ea-d168-4d94-a65a-30f616a17642.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkyMjgwNzcsIm5iZiI6MTcyOTIyNzc3NywicGF0aCI6Ii8xNzk1Mzk1OTgvMzc3Njk0OTIwLTg5YmU5M2VhLWQxNjgtNGQ5NC1hNjVhLTMwZjYxNmExNzY0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxOFQwNTAyNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWQxYzQxODU3Mzg2YThmYTgwOWZkMmZiNTc3Yzk5YzY4NTQ5NjFjNWM5YTE0NmZkNTcyNWMzZmRjNmUzMWMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Lr-vmVgm8GaW0oq6m4AwNvHdIYGTPXLahxu2C1Bqfh4)

If what I'm getting at here is right, the script is actually trying to load a simple _resource_ instead of a _scene_, so technically your code should be working - just not in the way you intended. Resources and nodes (which scenes are a type of the latter) are fundamentally different in how they work: a resource is essentially a basic object that can have properties, while nodes are objects that can have "children" to form a basic hierarchy system, commonly using resources to specify certain properties or have their own properties.

There's a particular class in Godot, `SceneTree`, that handles the hierarchy of nodes. It usually is referenced by using the `get_tree()` method from any node. It also has two methods/functions specifically for switching scenes that help simplify the process: `change_scene_to_file()` and `change_scene_to_packed()`. The former switches the scene to a specified file, which should be what you want, while the latter switches it to a specified `PackedScene`, which is essentially a copy of a specified scene within the project that can be loaded with either `preload()` or `load()`. `PackedScene`s, to my knowledge, are also used when you add a reusable scene file to the current scene's hierarchy - usually by the editor (in which you can drag the specified scene file into either a specific position on the viewport, or into the hierarchy to make it a child of another node), or by script, in which the `PackedScene` has to be regularly instantiated and then added in as a child to any node to actually make it have an effect on most of the game's elements. A `PackedScene` is essentially a resource due to it usually having a value that can be retrieved by `preload()` or `load()`, depending on the current scenario the `PackedScene` is used in.

I would instead write the twelfth line with that logic like this:

```gdscript
get_tree().change_scene_to_file("res://Scene/Presentation/CharacterSelect.tscn")
```

This gets the `SceneTree` currently running (which is referenced to by `get_tree()`) and calls the `change_scene_to_file()` method, with the parameter being the path to the character selection screen in the image you provided.

You can also use the `nextScene` property specified in your script in place of the file path, since that is specified in your script. In this case, we can instead use `change_scene_to_packed()` for this purpose, since the operation you use to retrieve the scene is used through `load()` with a path to a scene making it a `PackedScene`:

```gdscript
get_tree().change_scene_to_packed(nextScene)
```

TL;DR, the method used in your original script is for resources instead of a scene. To switch to a scene, use `get_tree().change_scene_to_file("<filepath>")` or `get_tree().change_scene_to_packed(<packedscene>)`, depending on your use case.
